File: docs/samples/rn222_demo.py
"""
Démonstration bout-en-bout : 1000 Bq de Rn-222, activités des descendants à
t=3600 s. Combine filiation.py (décomposition en branches), bateman.py
(calcul vectorisé), lara_client.py (accès LARA par étiquette).

VALIDÉ contre un outil de référence indépendant (calculateur officiel
Nucléide-Lara du LNHB) : écart < 0,005 % sur les 9 nucléides de la chaîne
à t=3600 s (voir docs/RECONCILIATION.md §11 pour le tableau complet).

Données réelles LARA (Po-218, Pb-214, Bi-214, Po-214) dans
docs/samples/lara_real/. Pour Rn-222, At-218, Rn-218, Tl-210, Pb-210 :
demi-vies reprises de l'en-tête de la référence reçue (pas de fichier LARA
récupéré directement pour ces 5-là dans cette session, mais valeurs
confirmées cohérentes par la validation elle-même).

Exécuter : python3 docs/samples/rn222_demo.py
"""
import sys
import os
import math
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "python"))

import numpy as np
from nuc.lara_client import LaraClient, FetchResult
from nuc.filiation import wholechain
from nuc.bateman import bateman

_SAMPLES = os.path.join(os.path.dirname(__file__), "lara_real")


def _read(name):
    with open(os.path.join(_SAMPLES, name), encoding="utf-8") as f:
        return f.read()


_LAMBDA_PB210 = math.log(2) / (22.23 * 365.25 * 86400)  # T1/2 = 22.23 a

DATA = {
    "Rn-222": (
        "Nuclide ; Rn-222\nDaughter(s) ; (alpha) ; Po-218 ; 100\n"
        "Half-life (s) ; 330325.0 ; 100\nDecay constant (1/s) ; 2.0983E-6 ; 0.0006E-6\n"
    ),
    "Po-218": _read("Po-218.lara.txt"),
    "At-218": (
        # 99.9% alpha -> Bi-214 (reconvergence avec la branche principale,
        # cf. docs/RECONCILIATION.md §8) ; 0.1% beta- -> Rn-218
        "Nuclide ; At-218\nDaughter(s) ; (alpha) ; Bi-214 ; 99.9 ; (B-) ; Rn-218 ; 0.1\n"
        "Half-life (s) ; 1.4 ; 0.1\nDecay constant (1/s) ; 4.951E-1 ; 0.35E-1\n"
    ),
    "Pb-214": _read("Pb-214.lara.txt"),
    "Rn-218": (
        "Nuclide ; Rn-218\nDaughter(s) ; (alpha) ; Po-214 ; 100\n"
        "Half-life (s) ; 0.036 ; 0.003\nDecay constant (1/s) ; 19.25 ; 1.6\n"
    ),
    "Bi-214": _read("Bi-214.lara.txt"),
    "Po-214": _read("Po-214.lara.txt"),
    "Tl-210": (
        "Nuclide ; Tl-210\nDaughter(s) ; (B-) ; Pb-210 ; 100\n"
        "Half-life (s) ; 78.0 ; 0.6\nDecay constant (1/s) ; 8.887E-3 ; 0.068E-3\n"
    ),
    "Pb-210": f"Nuclide ; Pb-210\nDecay constant (1/s) ; {_LAMBDA_PB210:.6e} ; 0\n",
}


def _fetch(url, timeout):
    for nuclide, text in DATA.items():
        if nuclide in url:
            return FetchResult(200, text)
    return FetchResult(404, "")


def main():
    client = LaraClient(_fetch, read_local_cache_fn=lambda n: "")
    branches = wholechain(client, "Rn-222")

    print(f"{len(branches)} branche(s) de filiation :")
    for b in branches:
        print("  " + " -> ".join(b.names))

    A0_Bq = 1000.0  # meme base que la reference recue
    lam_rn222 = client.decay_constant("Rn-222")[0]
    N0_rn222 = A0_Bq / lam_rn222

    t = 3600.0
    totals = {}
    lambdas_by_name = {}
    for b in branches:
        N0 = [N0_rn222 if i == 0 else 0.0 for i in range(len(b.names))]
        result = bateman(np.array(b.lambdas), np.array(b.ratios), np.array(N0), t)
        for name, n, c, lam in zip(b.names, result, b.conct, b.lambdas):
            lambdas_by_name[name] = lam
            if c:
                totals[name] = totals.get(name, 0.0) + n

    print(f"\nActivités à t={t:.0f} s (départ : {A0_Bq:.0f} Bq de Rn-222 pur) :")
    for name, n_atoms in totals.items():
        activity = n_atoms * lambdas_by_name[name]
        print(f"  {name:8s} : {activity:9.5f} Bq")


if __name__ == "__main__":
    main()

File: docs/samples/rn222_demo.m
function rn222_demo()
% RN222_DEMO Meme demonstration que rn222_demo.py, en Matlab : verifie la
% parite entre les deux portages (filiation.m/bateman.m vs filiation.py/
% bateman.py), tous deux valides independamment contre l'outil LNHB
% (docs/RECONCILIATION.md paragraphe 11).
    addpath(fullfile(fileparts(mfilename('fullpath')), '..', '..', 'matlab'));

    m = containers.Map();
    m('Rn-222') = sprintf(['Nuclide ; Rn-222\nDaughter(s) ; alpha ; Po-218 ; 100\n' ...
        'Decay constant (1/s) ; 2.0983E-6 ; 0.0006E-6\n']);
    m('Po-218') = fileread(fullfile(fileparts(mfilename('fullpath')), 'lara_real', 'Po-218.lara.txt'));
    m('At-218') = sprintf(['Nuclide ; At-218\nDaughter(s) ; alpha ; Bi-214 ; 99.9 ; B- ; Rn-218 ; 0.1\n' ...
        'Decay constant (1/s) ; 4.951E-1 ; 0.35E-1\n']);
    m('Pb-214') = fileread(fullfile(fileparts(mfilename('fullpath')), 'lara_real', 'Pb-214.lara.txt'));
    m('Rn-218') = sprintf(['Nuclide ; Rn-218\nDaughter(s) ; alpha ; Po-214 ; 100\n' ...
        'Decay constant (1/s) ; 19.25 ; 1.6\n']);
    m('Bi-214') = fileread(fullfile(fileparts(mfilename('fullpath')), 'lara_real', 'Bi-214.lara.txt'));
    m('Po-214') = fileread(fullfile(fileparts(mfilename('fullpath')), 'lara_real', 'Po-214.lara.txt'));
    m('Tl-210') = sprintf(['Nuclide ; Tl-210\nDaughter(s) ; B- ; Pb-210 ; 100\n' ...
        'Decay constant (1/s) ; 8.887E-3 ; 0.068E-3\n']);
    lam_pb210 = log(2) / (22.23*365.25*86400);
    m('Pb-210') = sprintf('Nuclide ; Pb-210\nDecay constant (1/s) ; %.6e ; 0\n', lam_pb210);

    keys_ = keys(m);
    function txt = fetch(nuclide)
        txt = '';
        for i = 1:length(keys_)
            if ~isempty(strfind(nuclide, keys_{i}))
                txt = m(keys_{i});
                return
            end
        end
    end
    client = lara_client_new(@fetch);

    branches = filiation(client, 'Rn-222');
    fprintf('%d branche(s) de filiation :\n', length(branches));
    for i = 1:length(branches)
        fprintf('  %s\n', strjoin(branches(i).names, ' -> '));
    end

    A0 = 1000.0;
    lam_rn222 = lara_decay_constant(client, 'Rn-222');
    N0_rn222 = A0 / lam_rn222;
    t = 3600.0;

    totals = containers.Map();
    lambdas_by_name = containers.Map();
    for i = 1:length(branches)
        b = branches(i);
        N0 = zeros(1, length(b.names)); N0(1) = N0_rn222;
        result = squeeze(bateman(b.lambdas, b.ratios, N0, t))';
        for j = 1:length(b.names)
            lambdas_by_name(b.names{j}) = b.lambdas(j);
            if b.conct(j)
                if isKey(totals, b.names{j})
                    totals(b.names{j}) = totals(b.names{j}) + result(j);
                else
                    totals(b.names{j}) = result(j);
                end
            end
        end
    end

    fprintf('\nActivités à t=%.0f s (départ : %.0f Bq de Rn-222 pur) :\n', t, A0);
    names_ = keys(totals);
    for i = 1:length(names_)
        n = names_{i};
        activity = totals(n) * lambdas_by_name(n);
        fprintf('  %-8s : %9.5f Bq\n', n, activity);
    end
end
